<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>gene position counts — gene_pos_counts • GARCOM</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="gene position counts — gene_pos_counts" />
<meta property="og:description" content="Function returns matrix with allelic counts per gene per individual for SNP and gene coordinates as inputs" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GARCOM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>gene position counts</h1>
    
    <div class="hidden name"><code>gene_pos_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Function returns matrix with allelic counts per gene per individual for SNP and gene coordinates as inputs</p>
    </div>

    <pre class="usage"><span class='fu'>gene_pos_counts</span>(<span class='no'>dt_gen</span>,<span class='no'>dt_snp</span>,<span class='no'>dt_gene</span>,<span class='kw'>keep_indiv</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>extract_SNP</span><span class='kw'>=</span><span class='kw'>NULL</span>,<span class='kw'>filter_gene</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>impute_missing</span><span class='kw'>=</span><span class='fl'>FALSE</span>,<span class='kw'>impute_method</span><span class='kw'>=</span><span class='st'>"mean"</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dt_gen</th>
      <td><p>a dataframe for genetic data that follows PLINK format (.raw)</p></td>
    </tr>
    <tr>
      <th>dt_snp</th>
      <td><p>a dataframe for SNP information with SNP BP as column names.</p></td>
    </tr>
    <tr>
      <th>dt_gene</th>
      <td><p>a dataframe for gene boundaries with CHR START END GENE as column names. Where CHR should be integer 1-22. START and END column should be integer. GENE column contains gene names</p></td>
    </tr>
    <tr>
      <th>keep_indiv</th>
      <td><p>an option to specify individuals to retain. Mutation counts will be provided for individuals provided in the list only. Default is all individuals.</p></td>
    </tr>
    <tr>
      <th>extract_SNP</th>
      <td><p>an option to specify SNPs for which mutation counts are needed. Mutation counts will be provided for SNPs included in the list only. Default is all SNPs.</p></td>
    </tr>
    <tr>
      <th>filter_gene</th>
      <td><p>an option to filter in Genes. Mutation counts will be provided for genes included in the list only. Default is all genes.</p></td>
    </tr>
    <tr>
      <th>impute_missing</th>
      <td><p>an option to impute missing genotypes. Default is FALSE.</p></td>
    </tr>
    <tr>
      <th>impute_method</th>
      <td><p>an option to specify method to specify imptuation method. Default method is impute to the mean. Alternatively imputation can be carried out by median. Function accepts method in quotes: "mean" or "median". Data are rounded to the second decimal places (e.g. 0.1234 will become 0.12.).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns an object of data.table class as an output with allelic gene counts within each sample where each row corresponds to gene and column to individual IDs from column second. The first column contains gene names.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Inputs needed are: recoded genetic data formatted in PLINK format, SNP name with BP (position) and gene name with START and END position. The first six columns of the input genetic data follow standard PLINK .raw format. Column names as FID, IID, PAT, MAT, SEX and PHENOTYPE followed by SNP information as recoded by the PLINK software. The function returns allelic counts per gene per sample (where each row represents a gene and each column represents an individual starting with the second column where first column contains gene information).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#Package provides sample data that are loaded with package loading. </span>
<span class='co'>#not RUN</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>recodedgen</span>) <span class='co'>#PLINK raw formatted data of 10 individiduals with 10 SNPs</span>

<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>genecoord</span>) <span class='co'>#gene coordinates with START, END, CHR and GENE names. </span>
<span class='co'>#Five genes with start and end genomic coordinates</span>

<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>snppos</span>) <span class='co'>#SNP and BP column names with SNP names and SNP genomic location in BP. </span>
<span class='co'>#10 SNPs with genomic location</span>

<span class='fu'>gene_pos_counts</span>(<span class='no'>recodedgen</span>, <span class='no'>snppos</span>, <span class='no'>genecoord</span>) <span class='co'>#run the function</span></div><div class='output co'>#&gt;     GENE IID_sample1 IID_sample2 IID_sample3 IID_sample4 IID_sample5
#&gt; 1: GENE1           1           0           0           0           0
#&gt; 2: GENE2           0           0           1           1           1
#&gt; 3: GENE3           0           0           0           0           0
#&gt;    IID_sample6 IID_sample7 IID_sample8 IID_sample9 IID_sample10
#&gt; 1:           0           0           0           0            0
#&gt; 2:           1           1           1           1            1
#&gt; 3:           0           0           0           0            1</div><div class='input'>
<span class='co'>#subset individuals</span>
<span class='fu'>gene_pos_counts</span>(<span class='no'>recodedgen</span>, <span class='no'>snppos</span>, <span class='no'>genecoord</span>,<span class='kw'>keep_indiv</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"IID_sample2"</span>,<span class='st'>"IID_sample4"</span>))</div><div class='output co'>#&gt;     GENE IID_sample2 IID_sample4
#&gt; 1: GENE2           0           1</div><div class='input'>
<span class='co'>#subset genes</span>
<span class='fu'>gene_pos_counts</span>(<span class='no'>recodedgen</span>,<span class='no'>snppos</span>,<span class='no'>genecoord</span>,<span class='kw'>filter_gene</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GENE1"</span>,<span class='st'>"GENE2"</span>))</div><div class='output co'>#&gt;     GENE IID_sample1 IID_sample2 IID_sample3 IID_sample4 IID_sample5
#&gt; 1: GENE1           1           0           0           0           0
#&gt; 2: GENE2           0           0           1           1           1
#&gt;    IID_sample6 IID_sample7 IID_sample8 IID_sample9 IID_sample10
#&gt; 1:           0           0           0           0            0
#&gt; 2:           1           1           1           1            1</div><div class='input'>
<span class='co'>#subset genes and individual iids</span>
<span class='fu'>gene_pos_counts</span>(<span class='no'>recodedgen</span>,<span class='no'>snppos</span>,<span class='no'>genecoord</span>,<span class='kw'>filter_gene</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GENE1"</span>,<span class='st'>"GENE2"</span>),
<span class='kw'>keep_indiv</span><span class='kw'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"IID_sample10"</span>,<span class='st'>"IID_sample4"</span>))</div><div class='output co'>#&gt;     GENE IID_sample10 IID_sample4
#&gt; 1: GENE2            1           1</div><div class='input'>
<span class='co'>##impute by mean</span>
<span class='fu'>gene_pos_counts</span>(<span class='no'>recodedgen</span>,<span class='no'>snppos</span>,<span class='no'>genecoord</span>,<span class='kw'>impute_missing</span><span class='kw'>=</span><span class='fl'>TRUE</span>,<span class='kw'>impute_method</span><span class='kw'>=</span><span class='st'>"mean"</span>)</div><div class='output co'>#&gt;     GENE IID_sample1 IID_sample2 IID_sample3 IID_sample4 IID_sample5
#&gt; 1: GENE1        1.00           0           0           0           0
#&gt; 2: GENE2        0.00           0           1           1           1
#&gt; 3: GENE3        0.11           0           0           0           0
#&gt;    IID_sample6 IID_sample7 IID_sample8 IID_sample9 IID_sample10
#&gt; 1:           0           0           0           0            0
#&gt; 2:           1           1           1           1            1
#&gt; 3:           0           0           0           0            1</div><div class='input'>
#end not RUN

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Sanjeev Sariya, Giuseppe Tosto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


